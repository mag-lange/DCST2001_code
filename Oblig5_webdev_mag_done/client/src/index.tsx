import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { Card, Row, Column, Form, Button } from './widgets';
import taskService, { Task } from './task-service';

class TaskList extends Component {
  tasks: Task[] = [];

  render() {
    return (
      <Card title="Tasks">
        {this.tasks.map((task) => (
          <Row key={task.id}>
            <Column>
              {task.id + '. ' + task.title + ' '}
              <input
                type="checkbox"
                defaultChecked={task.done}
                onChange={() => {
                  taskService.changeDonenotDone({
                    id: task.id,
                    title: task.title,
                    done: !task.done,
                  });
                  TaskList.instance()?.mounted();
                }}
              ></input>{' '}
              <Button.Danger
                onClick={() => {
                  taskService.delete(task.id);
                  TaskList.instance()?.mounted(); //Reload to show that the item has actually been deleted
                }}
              >
                x
              </Button.Danger>
            </Column>
          </Row>
        ))}
      </Card>
    );
  }

  mounted() {
    taskService.getAll().then((tasks) => (this.tasks = tasks));
  }
}

class TaskNew extends Component {
  title = '';

  render() {
    return (
      <Card title="New task">
        <Row>
          <Column width={1}>
            <Form.Label>Title:</Form.Label>
          </Column>
          <Column width={4}>
            <Form.Input
              type="text"
              value={this.title}
              onChange={(event) => (this.title = event.currentTarget.value)}
            />
          </Column>
        </Row>
        <Button.Success
          onClick={() => {
            taskService.create(this.title).then(() => {
              // Reloads the tasks in the Tasks component
              TaskList.instance()?.mounted(); // .? meaning: call TaskList.instance().mounted() if TaskList.instance() does not return null
              this.title = '';
            });
          }}
        >
          Create
        </Button.Success>
      </Card>
    );
  }
}

let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <>
      <TaskList />
      <TaskNew />
    </>,
  );
