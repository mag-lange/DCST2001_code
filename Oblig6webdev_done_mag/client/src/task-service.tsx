import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Task = {
  id: number;
  title: string;
  done: boolean;
  description: string;
};

class TaskService {
  /**
   * Get task with given id.
   */
  get(id: number) {
    return axios.get<Task>('/tasks/' + id).then((response) => response.data);
  }

  /**
   * Get all tasks.
   */
  getAll() {
    return axios.get<Task[]>('/tasks').then((response) => response.data);
  }

  /**
   * Create new task having the given title.
   *
   * Resolves the newly created task id.
   */
  create(title: string, description: string, done: boolean) {
    return axios
      .post<{ id: number }>('/tasks', { title: title, description: description, done: done })
      .then((response) => response.data.id);
  }

  delete(id: number) {
    return axios.delete<{ id: number }>('/tasks/' + id);
  }

  update(id: number, title: string, description: string, done: boolean) {
    return axios
      .put<void>("/tasks/" + id, { id: id, title: title, description: description, done: done })
      .then((response) => response.data);
  }
  
}

const taskService = new TaskService();
export default taskService;
