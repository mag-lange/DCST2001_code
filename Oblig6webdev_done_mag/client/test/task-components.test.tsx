import * as React from 'react';
import { TaskList, TaskDetails, TaskEdit } from '../src/task-components';
import { shallow } from 'enzyme';
import { Form, Button, Column, Row, Card, Alert } from '../src/widgets';
import { NavLink } from 'react-router-dom';


jest.mock('../src/task-service', () => {
  class TaskService {
    getAll() {
      return Promise.resolve([
        {
          id: 1,
          title: 'demo-task-title',
          description: 'demo-description',
          done: false,
        },
        { id: 2, title: 'Møt opp på forelesning', description: '', done: false },
        { id: 3, title: 'Gjør øving', description: '', done: false },
      ]);
    }

    get() {
      return Promise.resolve({
        id: 1,
        title: 'demo-task-title',
        description: 'demo-description',
        done: false,
      });
    }

    create() {
      return Promise.resolve(4);
    }

    update() {
      return Promise.resolve();
    }

    delete() {
      return Promise.resolve();
    }
  }

  return new TaskService();
});

describe('Testing task component', () => {
  test('TaskList correct initialization', (done) => {
    const wrapper = shallow(<TaskList />);

    // Wait for events to complete
    setTimeout(() => {
      expect(
        wrapper.containsAllMatchingElements([
          <NavLink to="/tasks/1">demo-task-title</NavLink>,
          <NavLink to="/tasks/2">Møt opp på forelesning</NavLink>,
          <NavLink to="/tasks/3">Gjør øving</NavLink>,
        ]),
      ).toEqual(true);
      done();
    });
  });

  test('Correct action after clicking on tasks', (done) => {
    const wrapper = shallow(<TaskList />);

    wrapper.find(Button.Success).simulate('click');
    expect(location.hash).toEqual('#/tasks/new'); 
    done();
  });


});

describe('Testing TaskDetails', () => {
  test('TaskDetails correct initialization', (done) => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update();

      console.log(wrapper.debug()); //From GPT

      
      expect(
        wrapper.containsAllMatchingElements([
          <Card title="Task">
            <Row>
              <Column width={2}>Title:</Column>
              <Column>demo-task-title</Column>
            </Row>
            <Row>
              <Column width={2}>Description:</Column>
              <Column>demo-description</Column>
            </Row>
            <Row>
              <Column width={2}>Done:</Column>
              <Column>
              {/* @ts-ignore*/}
                <Form.Checkbox checked={false} /> 
              </Column>
            </Row>
          </Card>,
          // @ts-ignore
          <Button.Success>Edit</Button.Success>,
        ]),
      ).toEqual(true);
      done();
    }, 0);
  });
});

describe('TaskDetails correct initialization (snapshot)', () => {
  test('Renders correctly', () => {
    const wrapper = shallow(<TaskDetails match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update();
      expect(wrapper).toMatchSnapshot();
    }, 0);
  });
});

describe('TaskEdit correct initialization (snapshot)', () => {
  test('TaskEdit correct initialization (snapshot)', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update();

     // @ts-ignore
      expect(wrapper).toMatchSnapshot('before-edit-click');

      wrapper.find(Button.Success).simulate('click');

  // @ts-ignore
      expect(wrapper).toMatchSnapshot('after-edit-click');

      expect(location.hash).toEqual('#/tasks');
      done();
    }, 0);
  });
});

describe('TaskEdit tests', () => {

test('TaskEdit renders correctly', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update(); // Ensure the component is updated

      
      expect(
        wrapper.containsAllMatchingElements([
          <Card title="Edit task">
             {/* @ts-ignore*/}
            <Form.Input value="demo-task-title" />,
            {/* @ts-ignore*/}
            <Form.Input value="demo-description" />,
            {/* @ts-ignore*/}
            <Form.Checkbox checked={false} />,
          </Card>,
          //@ts-ignore
          <Button.Success>Save</Button.Success>,
        ]),
      ).toEqual(false);
      done();
    }, 0);
  });


  test('TaskEdit finishing task', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update();

      wrapper.find(Form.Checkbox).simulate('change', { currentTarget: { checked: false } });
      expect(wrapper.find(Form.Checkbox).prop('checked')).toEqual(false);
      done();
    }, 0);
  });

  test('TaskEdit go to edit page', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update();

      wrapper.find(Button.Success).simulate('click');
      expect(location.hash).toEqual('#/tasks');
      done();
    }, 0);
  });

  test('TaskEdit check coloumn width', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update(); 

      expect(wrapper.containsMatchingElement(<Column width={2}>Title:</Column>)).toEqual(false);
      done();
    }, 0);
  });


  test('TaskEdit get into delete button', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);

    setTimeout(() => {
      wrapper.update(); 
      wrapper.find(Button.Danger).simulate('click');
      expect(location.hash).toEqual('#/tasks');
      done();
    }, 0);
  });

  test('TaskEdit press the save button', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);
    setTimeout(() => {
      wrapper.update(); 

      wrapper.find(Button.Success).simulate('click');
      expect(location.hash).toEqual('#/tasks');
      done();
    }, 0);
  });

  test('TaskEdit text area', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);
    setTimeout(() => {
      wrapper.update();
      //@ts-ignore
      expect(wrapper.containsMatchingElement(<Form.Textarea rows={10} />)).toEqual(true);
      done();
    }, 0);
  });

  test('TaskEdit editing title', (done) => {
    const wrapper = shallow(<TaskEdit match={{ params: { id: 1 } }} />);
    setTimeout(() => {
      wrapper.update();

      wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'edit' } });
      expect(wrapper.find(Form.Input).prop('value')).toEqual('edit');
      done();
    }, 0);
  });
});
