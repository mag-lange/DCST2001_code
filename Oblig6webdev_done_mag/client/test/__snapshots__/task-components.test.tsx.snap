// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskEdit correct initialization (snapshot) TaskEdit correct initialization (snapshot) 1`] = `
<Fragment>
  <Card
    title="Task"
  >
    <Row>
      <Column
        width={2}
      >
        Title:
      </Column>
      <Column>
        demo-task-title
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Description:
      </Column>
      <Column
        width={2}
      >
        demo-description
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          disabled={true}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <ButtonSuccess
    onClick={[Function]}
  >
    Edit
  </ButtonSuccess>
</Fragment>
`;

exports[`TaskEdit correct initialization (snapshot) TaskEdit correct initialization (snapshot): after-edit-click 1`] = `
<Fragment>
  <Card
    title="Edit task"
  >
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Title:
        </FormLabel>
      </Column>
      <Column>
        <FormInput
          onChange={[Function]}
          type="text"
          value="demo-task-title"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Description:
        </FormLabel>
      </Column>
      <Column>
        <FormTextarea
          onChange={[Function]}
          rows={10}
          value="demo-description"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <Row>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      right={true}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`TaskEdit correct initialization (snapshot) TaskEdit correct initialization (snapshot): before-edit-click 1`] = `
<Fragment>
  <Card
    title="Edit task"
  >
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Title:
        </FormLabel>
      </Column>
      <Column>
        <FormInput
          onChange={[Function]}
          type="text"
          value="demo-task-title"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Description:
        </FormLabel>
      </Column>
      <Column>
        <FormTextarea
          onChange={[Function]}
          rows={10}
          value="demo-description"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <Row>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      right={true}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`TaskEdit renders correctly with use of snapshot TaskEdit renders correctly with use of snapshot 1`] = `
<Fragment>
  <Card
    title="Task"
  >
    <Row>
      <Column
        width={2}
      >
        Title:
      </Column>
      <Column>
        demo-task-title
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Description:
      </Column>
      <Column
        width={2}
      >
        demo-description
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          disabled={true}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <ButtonSuccess
    onClick={[Function]}
  >
    Edit
  </ButtonSuccess>
</Fragment>
`;

exports[`TaskEdit renders correctly with use of snapshot TaskEdit renders correctly with use of snapshot: after-edit-click 1`] = `
<Fragment>
  <Card
    title="Edit task"
  >
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Title:
        </FormLabel>
      </Column>
      <Column>
        <FormInput
          onChange={[Function]}
          type="text"
          value="demo-task-title"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Description:
        </FormLabel>
      </Column>
      <Column>
        <FormTextarea
          onChange={[Function]}
          rows={10}
          value="demo-description"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <Row>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      right={true}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`TaskEdit renders correctly with use of snapshot TaskEdit renders correctly with use of snapshot: before-edit-click 1`] = `
<Fragment>
  <Card
    title="Edit task"
  >
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Title:
        </FormLabel>
      </Column>
      <Column>
        <FormInput
          onChange={[Function]}
          type="text"
          value="demo-task-title"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Description:
        </FormLabel>
      </Column>
      <Column>
        <FormTextarea
          onChange={[Function]}
          rows={10}
          value="demo-description"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <Row>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      right={true}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`TaskEdit save changes correctly: after-edit-click 1`] = `
<Fragment>
  <Card
    title="Edit task"
  >
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Title:
        </FormLabel>
      </Column>
      <Column>
        <FormInput
          onChange={[Function]}
          type="text"
          value="demo-task-title"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Description:
        </FormLabel>
      </Column>
      <Column>
        <FormTextarea
          onChange={[Function]}
          rows={10}
          value="demo-description"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <Row>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      right={true}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;

exports[`TaskEdit save changes correctly: before-edit-click 1`] = `
<Fragment>
  <Card
    title="Edit task"
  >
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Title:
        </FormLabel>
      </Column>
      <Column>
        <FormInput
          onChange={[Function]}
          type="text"
          value="demo-task-title"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        <FormLabel>
          Description:
        </FormLabel>
      </Column>
      <Column>
        <FormTextarea
          onChange={[Function]}
          rows={10}
          value="demo-description"
        />
      </Column>
    </Row>
    <Row>
      <Column
        width={2}
      >
        Done:
      </Column>
      <Column>
        <FormCheckbox
          checked={false}
          onChange={[Function]}
        />
      </Column>
    </Row>
  </Card>
  <Row>
    <Column>
      <ButtonSuccess
        onClick={[Function]}
      >
        Save
      </ButtonSuccess>
    </Column>
    <Column
      right={true}
    >
      <ButtonDanger
        onClick={[Function]}
      >
        Delete
      </ButtonDanger>
    </Column>
  </Row>
</Fragment>
`;
